File: producer-consumer/src/main/java/com/wardziniak/kafka/partitioner/CityPartitioner.java
package com.wardziniak.kafka.partitioner;

import com.wardziniak.kafka.model.Person;
import org.apache.kafka.clients.producer.Partitioner;
import org.apache.kafka.common.Cluster;

import java.util.Map;

/**
 * Created by wardziniak on 10.06.2018.
 */
public class CityPartitioner implements Partitioner {

    @Override
    public int partition(String topic, Object key, byte[] keyBytes, Object value, byte[] valueBytes, Cluster cluster) {
        if (value instanceof Person)
        {
            Person person = (Person)value;
            return person.getAddress().getCity().hashCode() % cluster.partitionCountForTopic(topic);
        }
        else
            return key.hashCode() % cluster.partitionCountForTopic(topic);
    }

    @Override
    public void close() {

    }

    @Override
    public void configure(Map<String, ?> configs) {

    }
}

File: producer-consumer/src/main/scala/com/wardziniak/kafka/scala/partitioner/CityPartitioner.scala
package com.wardziniak.kafka.scala.partitioner

import java.util

import com.wardziniak.kafka.utils.model.Person
import org.apache.kafka.clients.producer.Partitioner
import org.apache.kafka.common.Cluster

class CityPartitioner extends Partitioner {
  override def partition(topic: String, key: scala.Any, keyBytes: Array[Byte], value: scala.Any, valueBytes: Array[Byte], cluster: Cluster): Int =
    value match {
      case person: Person => person.address.city.hashCode % cluster.partitionCountForTopic(topic)
      case _ => key.hashCode % cluster.partitionCountForTopic(topic);
    }

  override def close(): Unit = {}

  override def configure(configs: util.Map[String, _]): Unit = {}
}
